src/kernel.cc

```diff
@@ -741,33 +783,101 @@ void quantum_kernel::gate(const Str &gname, UInt q0, UInt q1) {
     const Vec<UInt> &qubits,
     const Vec<UInt> &cregs,
     UInt duration,
-    Real angle
+    Real angle,
+    const Vec<UInt> &bregs,
+    cond_type_t gcond,
+    const Vec<UInt> &gcondregs
 ) {
-    /// @todo-rn: move these check to a platform-specific backend after qubits are initialized
+    QL_DOUT("gate:" <<" gname=" << gname <<" qubits=" << qubits <<" cregs=" << cregs <<" duration=" << duration <<" angle=" << angle <<" bregs=" << bregs <<" gcond=" << gcond <<" gcondregs=" << gcondregs);
+
     for (auto &qno : qubits) {
         if (qno >= qubit_count) {
             QL_FATAL("Number of qubits in platform: " << to_string(qubit_count) << ", specified qubit numbers out of range for gate: '" << gname << "' with qubits " << qubits);
         }
     }
-
-    for (auto & cno : cregs) {
+    for (auto &cno : cregs) {
         if (cno >= creg_count) {
             QL_FATAL("Out of range operand(s) for '" << gname << "' with cregs " << cregs);
         }
     }
+    for (auto &bno : bregs) {
+        if (bno >= breg_count) {
+            QL_FATAL("Out of range operand(s) for '" << gname << "' with bregs " << bregs);
+        }
+    }
+    if (!gate::is_valid_cond(gcond, gcondregs)) {
+        QL_FATAL("Condition " << gcond << " of '" << gname << "' incompatible with gcondregs " << gcondregs);
+    }
+    for (auto &cbno : gcondregs) {
+        if (cbno >= breg_count) {
+            QL_FATAL("Out of range condition operand(s) for '" << gname << "' with gcondregs " << gcondregs);
+        }
+    }
+    auto lqubits = qubits;
+    auto lcregs = cregs;
+    auto lbregs = bregs;
+    gate_add_implicits(gname, lqubits, lcregs, duration, angle, lbregs, gcond, gcondregs);
+    if (!gate_nonfatal(gname, lqubits, lcregs, duration, angle, lbregs, gcond, gcondregs)) {
+        QL_FATAL("Unknown gate '" << gname << "' with qubits " << lqubits);
+    }
+}
+
+/**
+ * preset condition to make all future created gates conditional gates with this condition
+ * preset ends when cleared: back to {cond_always, {}};
+ * useful in combination with higher-level gate creation interfaces
+ * that don't support adding a condition for conditional execution
+ */
+void quantum_kernel::gate_preset_condition(
+    cond_type_t gcond,
+    const utils::Vec<utils::UInt> &gcondregs
+) {
+    if (!gate::is_valid_cond(gcond, gcondregs)) {
+        QL_FATAL("Condition " << gcond << " of gate_preset_condition incompatible with gcondregs " << gcondregs);
+    }
+    QL_DOUT("Gate_preset_condition: setting condition=" << condition << " cond_operands=" << cond_operands);
+    condition = gcond;
+    cond_operands = gcondregs;
+}
+
+/**
+ * clear preset condition again
+ */
+void quantum_kernel::gate_clear_condition() {
+    gate_preset_condition(cond_always, {});
+}
 
-    if (!gate_nonfatal(gname, qubits, cregs, duration, angle)) {
-        QL_FATAL("Unknown gate '" << gname << "' with qubits " << qubits);
+/**
+ * add implicit parameters to gate to match IR requirements
```

Comment:
I don't understand the comment. Most parameters are unused (except for logging). Should probably be private in .h file

Reply:
Was meant as a place holder to expand gates without arguments as well.
It is really ugly to have this now implicit in the IR; see e.g. src/scheduler.cc when creating depgraph in Init().
And don't want to update the prototype each time we add an other translation.